File: sdks/python/src/untrace/untrace.py
"""Main Untrace SDK implementation."""

import logging
from typing import Any, Dict, List, Optional
from opentelemetry import trace
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
try:
    from opentelemetry.semantic_conventions.resource import ResourceAttributes
except ImportError:
    # Fallback for older versions
    class ResourceAttributes:
        SERVICE_NAME = "service.name"
        SERVICE_VERSION = "service.version"

from .context import UntraceContext
from .exporter import UntraceExporter
from .metrics import UntraceMetrics
from .tracer import UntraceTracer
from .types import UntraceConfig
from .instrumentation import instrument_all, get_supported_providers


logger = logging.getLogger(__name__)

# Global state management
_global_state = {
    "instance": None,
    "is_initialized": False,
}


class Untrace:
    """Main Untrace SDK class."""

    def __init__(self, config: UntraceConfig):
        """Initialize the Untrace SDK.

        Args:
            config: Untrace configuration
        """
        self.config = config

        # Set up logging if debug is enabled
        if config.debug:
            logging.basicConfig(level=logging.DEBUG)
            logger.info("[Untrace] Initializing SDK with debug enabled")

        # Create resource with service information
        resource_attributes = {
            ResourceAttributes.SERVICE_NAME: "untrace-app",
            ResourceAttributes.SERVICE_VERSION: config.version,
            **config.resource_attributes,
        }

        resource = Resource.create(resource_attributes)

        # Create exporter
        exporter = UntraceExporter(config)

        # Create batch span processor
        batch_span_processor = BatchSpanProcessor(
            exporter,
            max_export_batch_size=config.max_batch_size,
            max_queue_size=max(config.max_batch_size * 2, 10),
            export_timeout_millis=config.export_interval_ms,
        )

        # Create tracer provider
        self._provider = TracerProvider(
            resource=resource,
            active_span_processor=batch_span_processor,
        )

        # Register provider
        trace.set_tracer_provider(self._provider)

        # Initialize components
        self._tracer = UntraceTracer(
            self._provider.get_tracer("untrace-app", config.version)
        )
        self._metrics = UntraceMetrics()
        self._context = UntraceContext()

        # Auto-instrument if not disabled
        if not config.disable_auto_instrumentation:
            self._auto_instrument()

        # Store instance
        _global_state["instance"] = self

        if config.debug:
            logger.info("[Untrace] SDK initialized successfully")

    def _auto_instrument(self) -> None:
        """Auto-instrument supported providers."""
        providers = self.config.providers

        # Get all supported providers
        supported_providers = get_supported_providers()

        # Default providers if not specified
        default_providers = ["openai", "anthropic", "google", "microsoft", "aws", "cohere"]

        providers_to_instrument = (
            supported_providers if "all" in providers
            else [p for p in providers if p in supported_providers]
        )

        if self.config.debug:
            logger.info(f"[Untrace] Auto-instrumenting providers: {providers_to_instrument}")

        # Instrument all providers at once
        try:
            instrument_all(self._tracer, self._metrics, self._context, providers_to_instrument)
        except Exception as error:
            if self.config.debug:
                logger.warning(f"[Untrace] Failed to instrument providers: {error}")

    def get_tracer(self) -> UntraceTracer:
        """Get the tracer instance.

        Returns:
            UntraceTracer instance
        """
        return self._tracer

    def get_metrics(self) -> UntraceMetrics:
        """Get the metrics instance.

        Returns:
            UntraceMetrics instance
        """
        return self._metrics

    def get_context(self) -> UntraceContext:
        """Get the context instance.

        Returns:
            UntraceContext instance
        """
        return self._context

    def instrument(self, provider_name: str, module: Any) -> Any:
        """Manually instrument a provider.

        Args:
            provider_name: Name of the provider
            module: Module to instrument

        Returns:
            Instrumented module

        Raises:
            ValueError: If provider is not supported
        """
        if provider_name not in ["openai"]:
            raise ValueError(f"Unknown provider: {provider_name}")

        # TODO: Implement provider-specific instrumentation
        if self.config.debug:
            logger.info(f"[Untrace] Manually instrumenting {provider_name}")

        return module

    async def flush(self) -> None:
        """Flush all pending spans."""
        if self.config.debug:
            logger.info("[Untrace] Flushing spans...")

        self._provider.force_flush()

        if self.config.debug:
            logger.info("[Untrace] Flush completed")

    async def shutdown(self) -> None:
        """Shutdown the SDK."""
        if self.config.debug:
            logger.info("[Untrace] Shutting down SDK...")

        # Shutdown provider
        self._provider.shutdown()

        # Clear instance and reset singleton state
        if _global_state["instance"] == self:
            _global_state["instance"] = None
            _global_state["is_initialized"] = False

        if self.config.debug:
            logger.info("[Untrace] SDK shutdown complete")

    @staticmethod
    def get_instance() -> Optional["Untrace"]:
        """Get the current Untrace instance.

        Returns:
            Current Untrace instance or None
        """
        return _global_state["instance"]


def init(config: UntraceConfig) -> Untrace:
    """Initialize the Untrace SDK.

    Args:
        config: Untrace configuration

    Returns:
        Untrace instance

    Raises:
        RuntimeError: If SDK is already initialized
    """
    if _global_state["is_initialized"] and _global_state["instance"]:
        if config.debug:
            logger.info("[Untrace] SDK already initialized. Returning existing instance.")
        return _global_state["instance"]

    if _global_state["is_initialized"]:
        raise RuntimeError("Untrace SDK is already initialized but instance is null.")

    _global_state["is_initialized"] = True
    _global_state["instance"] = Untrace(config)
    return _global_state["instance"]


def get_untrace() -> Untrace:
    """Get the current Untrace instance.

    Returns:
        Current Untrace instance

    Raises:
        RuntimeError: If SDK is not initialized
    """
    if not _global_state["instance"]:
        raise RuntimeError("Untrace SDK not initialized. Call init() first.")
    return _global_state["instance"]

File: sdks/csharp/src/Untrace/Untrace.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using OpenTelemetry;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;
using System.Diagnostics;

namespace Untrace;

/// <summary>
/// Main Untrace SDK class for OpenTelemetry-based observability
/// </summary>
public class Untrace : IDisposable
{
    private readonly TracerProvider _tracerProvider;
    private readonly UntraceConfig _config;
    private readonly ILogger<Untrace>? _logger;
    private readonly ActivitySource _activitySource;
    private bool _disposed = false;

    /// <summary>
    /// Initializes a new instance of the Untrace class
    /// </summary>
    /// <param name="config">Configuration options</param>
    /// <param name="logger">Optional logger</param>
    public Untrace(UntraceConfig config, ILogger<Untrace>? logger = null)
    {
        _config = config ?? throw new ArgumentNullException(nameof(config));
        _logger = logger;

        if (string.IsNullOrWhiteSpace(_config.ApiKey))
        {
            throw new ArgumentException("API key is required", nameof(config));
        }

        _activitySource = new ActivitySource(_config.ServiceName, _config.Version);

        // Create resource with service information
        var resourceBuilder = ResourceBuilder.CreateDefault()
            .AddService(_config.ServiceName, _config.Version)
            .AddAttributes(new Dictionary<string, object>
            {
                ["service.environment"] = _config.Environment,
                ["service.namespace"] = "untrace",
            });

        // Add custom resource attributes
        foreach (var attribute in _config.ResourceAttributes)
        {
            resourceBuilder.AddAttributes(new Dictionary<string, object> { [attribute.Key] = attribute.Value });
        }

        // Create tracer provider
        _tracerProvider = Sdk.CreateTracerProviderBuilder()
            .SetResourceBuilder(resourceBuilder)
            .SetSampler(new TraceIdRatioBasedSampler(_config.SamplingRate))
            .AddSource(_config.ServiceName)
            .Build();

        _logger?.LogInformation("Untrace SDK initialized for service: {ServiceName}", _config.ServiceName);
    }

    /// <summary>
    /// Get the activity source for creating activities
    /// </summary>
    public ActivitySource ActivitySource => _activitySource;

    /// <summary>
    /// Create a new activity for tracing
    /// </summary>
    /// <param name="name">Activity name</param>
    /// <param name="kind">Activity kind</param>
    /// <param name="attributes">Initial attributes</param>
    /// <returns>Activity instance</returns>
    public Activity? StartActivity(
        string name,
        ActivityKind kind = ActivityKind.Internal,
        Dictionary<string, object>? attributes = null)
    {
        var activity = _activitySource.StartActivity(name, kind);

        if (activity != null && attributes != null)
        {
            foreach (var attribute in attributes)
            {
                activity.SetTag(attribute.Key, attribute.Value);
            }
        }

        return activity;
    }

    /// <summary>
    /// Create an LLM activity for tracing LLM operations
    /// </summary>
    /// <param name="operation">LLM operation type</param>
    /// <param name="provider">Provider name</param>
    /// <param name="model">Model name</param>
    /// <param name="attributes">Additional attributes</param>
    /// <returns>Activity instance</returns>
    public Activity? StartLLMActivity(
        string operation,
        string provider,
        string model,
        Dictionary<string, object>? attributes = null)
    {
        var activityAttributes = new Dictionary<string, object>
        {
            ["llm.operation"] = operation,
            ["llm.provider"] = provider,
            ["llm.model"] = model,
        };

        if (attributes != null)
        {
            foreach (var attribute in attributes)
            {
                activityAttributes[attribute.Key] = attribute.Value;
            }
        }

        return StartActivity($"llm.{operation}", ActivityKind.Client, activityAttributes);
    }

    /// <summary>
    /// Record token usage metrics
    /// </summary>
    /// <param name="tokenUsage">Token usage information</param>
    public void RecordTokenUsage(TokenUsage tokenUsage)
    {
        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetTag("llm.prompt.tokens", tokenUsage.PromptTokens);
            activity.SetTag("llm.completion.tokens", tokenUsage.CompletionTokens);
            activity.SetTag("llm.total.tokens", tokenUsage.TotalTokens);
            activity.SetTag("llm.model", tokenUsage.Model);
            activity.SetTag("llm.provider", tokenUsage.Provider);
        }
    }

    /// <summary>
    /// Record cost metrics
    /// </summary>
    /// <param name="cost">Cost information</param>
    public void RecordCost(Cost cost)
    {
        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetTag("llm.cost.prompt", cost.Prompt);
            activity.SetTag("llm.cost.completion", cost.Completion);
            activity.SetTag("llm.cost.total", cost.Total);
            activity.SetTag("llm.model", cost.Model);
            activity.SetTag("llm.provider", cost.Provider);
        }
    }

    /// <summary>
    /// Record an exception in the current activity
    /// </summary>
    /// <param name="exception">Exception to record</param>
    public void RecordException(Exception exception)
    {
        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetStatus(ActivityStatusCode.Error, exception.Message);
            activity.SetTag("error", true);
            activity.SetTag("error.message", exception.Message);
            activity.SetTag("error.type", exception.GetType().Name);
        }
    }

    /// <summary>
    /// Flush all pending spans
    /// </summary>
    public void Flush()
    {
        _tracerProvider.ForceFlush();
        _logger?.LogDebug("Flushed all pending spans");
    }

    /// <summary>
    /// Dispose the SDK
    /// </summary>
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    /// <summary>
    /// Dispose the SDK
    /// </summary>
    /// <param name="disposing">Whether to dispose managed resources</param>
    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed && disposing)
        {
            _tracerProvider?.Dispose();
            _activitySource?.Dispose();
            _disposed = true;
            _logger?.LogInformation("Untrace SDK disposed");
        }
    }
}

/// <summary>
/// Static class for initializing the Untrace SDK
/// </summary>
public static class UntraceSdk
{
    private static Untrace? _instance;
    private static readonly object _lock = new();

    /// <summary>
    /// Initialize the Untrace SDK
    /// </summary>
    /// <param name="config">Configuration options</param>
    /// <param name="logger">Optional logger</param>
    /// <returns>Untrace instance</returns>
    public static Untrace Init(UntraceConfig config, ILogger<Untrace>? logger = null)
    {
        lock (_lock)
        {
            if (_instance != null)
            {
                logger?.LogWarning("Untrace SDK already initialized. Returning existing instance.");
                return _instance;
            }

            _instance = new Untrace(config, logger);
            return _instance;
        }
    }

    /// <summary>
    /// Get the current Untrace instance
    /// </summary>
    /// <returns>Untrace instance</returns>
    /// <exception cref="InvalidOperationException">Thrown when SDK is not initialized</exception>
    public static Untrace GetInstance()
    {
        lock (_lock)
        {
            if (_instance == null)
            {
                throw new InvalidOperationException("Untrace SDK not initialized. Call Init() first.");
            }
            return _instance;
        }
    }

    /// <summary>
    /// Check if the SDK is initialized
    /// </summary>
    /// <returns>True if initialized, false otherwise</returns>
    public static bool IsInitialized()
    {
        lock (_lock)
        {
            return _instance != null;
        }
    }

    /// <summary>
    /// Shutdown the SDK
    /// </summary>
    public static void Shutdown()
    {
        lock (_lock)
        {
            _instance?.Dispose();
            _instance = null;
        }
    }
}
